src/main/scala/org/constellation/rewards/RewardsManager.scala

```diff
@@ -0,0 +1,219 @@
+package org.constellation.rewards
+
+import cats.effect._
+import cats.effect.concurrent.Ref
+import cats.implicits._
+import org.constellation.checkpoint.CheckpointService
+import org.constellation.consensus.Snapshot
+import org.constellation.domain.observation.Observation
+import org.constellation.p2p.PeerNotification
+import org.constellation.primitives.Schema.{AddressCacheData, CheckpointEdge}
+import org.constellation.primitives.{ChannelMessage, Transaction}
+import org.constellation.rewards.RewardsManager.rewardDuringEpoch
+import org.constellation.storage.{AddressService, SnapshotBroadcastService}
+import org.constellation.trust.TrustEdge
+
+class RewardsManager[F[_]: Concurrent](
+  eigenTrust: EigenTrust[F],
+  checkpointService: CheckpointService[F],
+  addressService: AddressService[F],
+  snapshotBroadcastService: SnapshotBroadcastService[F]
+) {
+  final val rewards: Ref[F, Map[String, Double]] = Ref.unsafe(Map.empty)
+
+  def updateBySnapshot(snapshot: Snapshot, snapshotHeight: Int): F[Unit] =
+    for {
+      observations <- observationsFromSnapshot(snapshot)
+      _ <- eigenTrust.retrain(observations)
+      trustMap <- eigenTrust.getTrustForAddressHashes
+
+      weightContributions = weightByTrust(trustMap) _ >>> weightByEpoch(snapshotHeight) >>> weightByStardust
+      contributions = calculateContributions(trustMap.keySet.toSeq)
+      distribution = weightContributions(contributions)
+
+      _ <- updateRewards(distribution)
+      _ <- updateAddressBalances(distribution)
+    } yield ()
+
+  def weightByStardust: Map[String, Double] => Map[String, Double] =
```

Comment:
Do we need this method? Or could we just use weightByStardust on the object?

Reply:
We can call `Stardust.weightByStardust` directly. I just did an aliasing for readability.